extensions/amp-form/0.1/form-validators.js

```diff
@@ -0,0 +1,370 @@
+   */
+  reportInput(input) {
+    const invalidType = getInvalidType(input);
+    if (invalidType) {
+      this.showValidationFor(input, invalidType);
+    }
+  }
+
+  /**
+   * Hides all validation messages.
+   */
+  hideAllValidations() {
+    const inputs = this.form.querySelectorAll('input,select,textarea');
+    for (let i = 0; i < inputs.length; i++) {
+      this.hideValidationFor(inputs[i]);
+    }
+  }
+
+  /**
+   * @param {!Element} input
+   * @param {!string} invalidType
+   * @return {?Element}
+   */
+  getValidationFor(input, invalidType) {
+    if (!input.id) {
+      return null;
+    }
+    const selector = `[visible-when-invalid=${invalidType}]` +
+        `[validation-for=${input.id}]`;
+    if (!this.inputValidationsDict_[selector]) {
+      this.inputValidationsDict_[selector] = this.doc.querySelector(selector);
+    }
+    return this.inputValidationsDict_[selector];
+  }
+
+  /**
+   * @param {!Element} input
+   * @param {string} invalidType
+   */
+  showValidationFor(input, invalidType) {
+    const validation = this.getValidationFor(input, invalidType);
+    if (!validation) {
+      return;
+    }
+
+    if (!validation.textContent.trim()) {
+      validation.textContent = input.validationMessage;
+    }
+
+    input.setAttribute('aria-invalid', 'true');
+    validation.classList.add('visible');
+    this.inputVisibleValidationDict_[input.id] = validation;
+  }
+
+  /**
+   * @param {!Element} input
+   */
+  hideValidationFor(input) {
+    const visibleValidation = this.getVisibleValidationFor(input);
+    if (!visibleValidation) {
+      return;
+    }
+
+    input.removeAttribute('aria-invalid');
+    visibleValidation.classList.remove('visible');
+    delete this.inputVisibleValidationDict_[input.id];
+  }
+
+  /**
+   * @param {!Element} input
+   * @return {?Element}
+   */
+  getVisibleValidationFor(input) {
+    if (!input.id) {
+      return null;
+    }
+    return this.inputVisibleValidationDict_[input.id];
+  }
+
+  /**
+   * Whether an input should validate after an interaction.
+   * @param {!Element} input
+   * @return {boolean}
+   */
+  shouldValidateOnInteraction(input) {
+    return !!this.getVisibleValidationFor(input);
+  }
+
+  /**
+   * @param {!Event} event
+   */
+  onInteraction(event) {
+    const input = /** @type {!Element} */ (event.target);
+    const shouldValidate = this.shouldValidateOnInteraction(input);
+    this.hideValidationFor(input);
+    if (shouldValidate && !event.target.checkValidity()) {
+      this.reportInput(input);
+    }
+  }
+
+  /** @override */
+  onBlur(event) {
+    this.onInteraction(event);
+  }
+
+  /** @override */
+  onInput(event) {
+    this.onInteraction(event);
+  }
+}
+
+
+class ShowFirstOnSubmitValidator extends AbstractCustomValidator {
+
+  /** @override */
+  report() {
+    this.hideAllValidations();
+    const inputs = this.form.querySelectorAll('input,select,textarea');
+    for (let i = 0; i < inputs.length; i++) {
```

Comment:
Do we want to define an iteration function like `forEachInput`? I feel with the `break` statements in some of these it makes that difficult. How would we stop iterations? I am cool with the repetition in the for loops anyhow.